### PR TITLE
S3 main registration

### PR DESCRIPTION
**Pull request by patrickkenyon**
S3 tasks:
- Built login page
- Created user table
- Validated email
- Added emails to database on login
- Created redirects if not logged in

Please review all files (minus css, node etc.)

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+var gulp = require('gulp')
+var sass = require('gulp-sass')
+
+gulp.task('sass', function() {
+    return gulp.src(['scss/pokestyles.scss'])
```
Comment (hoptond) [inline gulpfile.js:5]: The file this links to still has hotpink in it.

File: src/ErrorLog.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+class ErrorLog
+{
+    private const ERROR_LOG = 'errorLog.txt';
+
+    /**
+     * This logs messages to the error log
+     * @param string $error Message can be user defined or accessed from Exception object
+     */
+    public static function log(string $errorMessage) {
```
Comment (hoptond) [inline src/ErrorLog.php]: PSR2 says method and class opening braces should go on the next line

https://www.php-fig.org/psr/psr-2/

File: src/ValueObjects/Email.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace Pokedex\ValueObjects;
+
+use Pokedex\AbstractEmail;
+use Pokedex\Exceptions\EmailException;
+
+class Email extends AbstractEmail {
```
Comment (hoptond) [inline src/ValueObjects/Email.php]: see above

File: src/ValueObjects/Email.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    private $email;
+
+    /**
+     * Email constructor that validates the user email address
+     * @param string $email user defined email address
+     * @throws EmailException
+     */
+    public function __construct(string $email) {
```
Comment (hoptond) [inline src/ValueObjects/Email.php]: see above

File: src/DbConnection.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    public function __construct()
+    {
+        $this->db =  new PDO('mysql:dbname=pokedex_hb;host=127.0.0.1','root');
+    }
+
+    /**
+     * Function to get the $db property
+     * @return PDO connection to pokedex_hb database
+     */
+    public function getDB() {
```
Comment (hoptond) [inline src/DbConnection.php]: see above

**Review by hoptond — CHANGES_REQUESTED** [review]

File: gulpfile.js (hunk 1 above)
Comment (patrickkenyon) [reply gulpfile.js:5]: index.css has no corresponding scss file and so isn't actually implemented in this story. I think it has somehow found it's way into the project from later branches.

File: src/ErrorLog.php (hunk 2 above)
Comment (patrickkenyon) [reply src/ErrorLog.php]: Changed in 145a07c

File: src/ValueObjects/Email.php (hunk 3 above)
Comment (patrickkenyon) [reply src/ValueObjects/Email.php]: Changed in adda5e3

File: src/ValueObjects/Email.php (hunk 4 above)
Comment (patrickkenyon) [reply src/ValueObjects/Email.php]: Changed in a697038

File: src/DbConnection.php (hunk 5 above)
Comment (patrickkenyon) [reply src/DbConnection.php]: Changed in 832db0b

**Review by hoptond — DISMISSED** [review]

File: emailValidation.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+require_once ('vendor/autoload.php');
+
+use Pokedex\ValueObjects\Email;
+use Pokedex\Exceptions\EmailException;
+use Pokedex\User;
+use Pokedex\DbConnection;
+use Pokedex\ErrorLog;
+
+$userEmail = $_POST['email'] ?? FALSE; //$_POST['tbc'] needs to be replaced with actual input name
```
Comment (mporam) [inline emailValidation.php]: remove comment

File: index.php [hunk 7]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+session_start();
+
+if(!$_SESSION['loggedIn']) {
```
Comment (mporam) [inline index.php]: `if (`

File: login.php [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    <h1>POKÈDEX</h1>
+</header>
+<main>
+    <form method="post" action="emailValidation.php">
+        <input class="email" name="email" id="email" alt="email" type="email" placeholder=" Email...">
+        <input class="loginbutton" type="submit" value="Login">
+    </form>
+    <p class="login-error">
+        <?php
+        if(isset($_GET['error'])) {
```
Comment (mporam) [inline login.php]: `if (`

Comment (mporam) [inline login.php]: I think you could put this into a static method in a class, just to tidy up this file

File: scss/pokestyles.scss [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  top: 50%;
+  height: 70%;
+}
+
+main {
+  margin: 0 auto;
+  width: 400px;
+}
+
+label {
```
Comment (mporam) [inline scss/pokestyles.scss]: why? you have no label tags

File: src/DbConnection.php [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+class DbConnection
+{
+    private $db;
+    /**
+     * DbConnection constructor creates new connection to the pokedex_hb database
+     */
+    public function __construct()
+    {
+        $this->db =  new PDO('mysql:dbname=pokedex_hb;host=127.0.0.1','root');
```
Comment (mporam) [inline src/DbConnection.php:14]: its common practise although not required, to put the DB name, host and username into constants in this class

File: src/Exceptions/EmailException.php [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<?php
+
+namespace Pokedex\Exceptions;
+
+class EmailException extends \Exception {
```
Comment (mporam) [inline src/Exceptions/EmailException.php]: drop `{` onto next line

File: src/User.php [hunk 12]
```diff
@@ -0,0 +1,27 @@
+<?php
+namespace Pokedex;
+
```
Comment (mporam) [inline src/User.php:3]: extra whitespace

File: src/AbstractEmail.php [hunk 13]
```diff
@@ -0,0 +1,6 @@
+<?php
+namespace Pokedex;
+
+abstract class AbstractEmail
```
Comment (mporam) [inline src/AbstractEmail.php]: its common to put abstracts into an abstract folder

File: src/User.php [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    /**
+     * User constructor that checks if email address exists in the database.  Adds email address if it does not exist.
+     * @param AbstractEmail $email the validated email address from the user
+     * @param resource $db the database connection
+     */
+    public function __construct(AbstractEmail $email, \PDO $db)
+    {
+        $this->email = $email;
+
+        $stmt = $db->prepare('SELECT `id`, `email` FROM `users` WHERE `email` = :email;');
```
Comment (mporam) [inline src/User.php]: no try catch, what happens if this query fails?

File: src/ValueObjects/Email.php [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    private $email;
+
+    /**
+     * Email constructor that validates the user email address
+     * @param string $email user defined email address
+     * @throws EmailException
+     */
+    public function __construct(string $email)
+    {
+        if(!filter_var($email, FILTER_VALIDATE_EMAIL)) {
```
Comment (mporam) [inline src/ValueObjects/Email.php]: `if (`

**Review by mporam — CHANGES_REQUESTED** [review]

File: emailValidation.php (hunk 6 above)
Comment (imogenhallett) [reply emailValidation.php]: Done in d19ce09

File: index.php (hunk 7 above)
Comment (patrickkenyon) [reply index.php]: Changed in ec9ae2d

File: login.php (hunk 8 above)
Comment (patrickkenyon) [reply login.php]: Changed in ec9ae2d

Comment (patrickkenyon) [reply login.php]: Added new static class in 9c5ad26

File: scss/pokestyles.scss (hunk 9 above)
Comment (patrickkenyon) [reply scss/pokestyles.scss]: Removed the unused label styling in 2a810b2

File: src/Exceptions/EmailException.php (hunk 11 above)
Comment (patrickkenyon) [reply src/Exceptions/EmailException.php]: Changed in 4fcbdc5

File: src/User.php (hunk 12 above)
Comment (patrickkenyon) [reply src/User.php:3]: Removed in 2a16239

File: src/AbstractEmail.php (hunk 13 above)
Comment (patrickkenyon) [reply src/AbstractEmail.php]: Created in ed07096

File: src/User.php (hunk 14 above)
Comment (patrickkenyon) [reply src/User.php]: Changed in 4df0906

File: src/ValueObjects/Email.php (hunk 15 above)
Comment (patrickkenyon) [reply src/ValueObjects/Email.php]: Changed in b782d28

File: src/ViewHelpers/LoginViewHelper.php [hunk 16]
```diff
@@ -0,0 +1,25 @@
+<?php
+namespace Pokedex\ViewHelpers;
+
+
+class LoginViewHelper
+{
+    public static function outputError($get)
```
Comment (mporam) [inline src/ViewHelpers/LoginViewHelper.php]: no unit test

**Review by mporam — CHANGES_REQUESTED** [review]

File: src/ViewHelpers/LoginViewHelper.php (hunk 16 above)
Comment (patrickkenyon) [reply src/ViewHelpers/LoginViewHelper.php]: Written in d78ff23

**Comment by patrickkenyon** [comment]
Also added a default error message in a58bf9b

File: test/src/ViewHelpers/LoginViewHelper.php [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+require_once ('../../../src/ViewHelpers/LoginViewHelper.php');
+
+class TestLoginViewHelper extends TestCase
+{
+
+    public $get = [
+        'error' => 1
+    ];
+
+    public function testOutputErrorSuccess() {
```
Comment (mporam) [inline test/src/ViewHelpers/LoginViewHelper.php:10]: not testing all scenarios

File: src/ViewHelpers/LoginViewHelper.php [hunk 18]
```diff
@@ -4,19 +4,22 @@
...
                     break;
                 case 2:
-                    return 'An unknown error has occurred';
+                    return 'An unknown error has occurred. Please try again.';
                     break;
                 case 3:
                     return 'An email address is required';
                     break;
+                default:
+                    return 'An unknown error has occurred. Please try again.';
```
Comment (mporam) [inline src/ViewHelpers/LoginViewHelper.php:19]: this is not DRY

**Review by mporam — CHANGES_REQUESTED** [review]

File: test/src/ViewHelpers/LoginViewHelper.php (hunk 17 above)
Comment (patrickkenyon) [reply test/src/ViewHelpers/LoginViewHelper.php:10]: Changed in 2444e6d

File: src/ViewHelpers/LoginViewHelper.php (hunk 18 above)
Comment (patrickkenyon) [reply src/ViewHelpers/LoginViewHelper.php:19]: Desiccated in 40450ae

**Review by mporam — APPROVED** [review]

**Pull request merged**